cache/cache.go

```diff
@@ -135,10 +203,19 @@ func (dc *directoryCache) Add(blobHash string, p []byte) {
 			fmt.Printf("Warning: could not create a cache file at %q: %v\n", c, err)
 			return
 		}
-		defer f.Close()
-		if n, err := f.Write(p); err != nil || n != len(p) {
-			fmt.Printf("Warning: failed to write cache: %d(wrote)/%d(expected): %v\n",
-				n, len(p), err)
+		want := b2.Len()
+		if _, err := io.CopyN(f, b2, int64(want)); err != nil {
+			// cleanup
+			f.Close()
+			os.Remove(c)
+			fmt.Printf("Warning: failed to write cache: %v\n", err)
+		}
+
+		// Cache the opened file for future use. If "direct" option is specified, this
+		// won't be done. This option is useful for priventing file cache from being
+		// polluted by data that won't be accessed immediately.
+		if opt.direct || !dc.fileCache.add(key, f) {
+			f.Close()
```

Comment:
Will this f be deleted at some point in the future? If not, this fd will remain open until the layer is unmounted.
This equals to fd leakage when it comes to long-running program.

Reply:
The reference count for this object is decreased when `done()` callback is called.

https://github.com/ktock/stargz-snapshotter/blob/33f7d6fed1279b668af21e15dbf1573451c88221/cache/cache.go#L132

and if the ref count becomes 0 (means nobody refers to this object) by this operation, pre-defined finalization (i.e. closing the file) is invoked.

https://github.com/ktock/stargz-snapshotter/blob/33f7d6fed1279b668af21e15dbf1573451c88221/cache/cache.go#L315-L318

https://github.com/ktock/stargz-snapshotter/blob/33f7d6fed1279b668af21e15dbf1573451c88221/cache/cache.go#L92-L94